sdk/metric/integrator/simple/simple.go

```diff
@@ -151,9 +151,20 @@ func (b *Integrator) Process(accum export.Accumulation) error {
 	sameCollection := b.state.finishedCollection == value.updated
 	value.updated = b.state.finishedCollection
 
-	// An existing value will be found for some stateKey when:
-	// (a) stateful aggregation is being used
-	// (b) multiple accumulators are being used.
+	// At this point in the code, we have located an existing
+	// value for some stateKey.  This can be because:
```

Comment:
So the values are only deleted if they were not updated in the just-ran collection cycle. This purges any metrics that are no longer being updated from memory. We used to keep for 2 (maybe 3?) stale collection cycles and the logic was significantly more complex and it wasn't obvious that 2 was better than 1.

Reply:
The Accumulator is still keeping data for one extra cycle. This avoids shrinking and growing the map repeatedly in the Accumulator. I took the simpler approach here, because this code is already pretty complicated and it could be done as a future optimization. If we leave the non-stateful entries, the integrator's memory will grow without bound.

(I think this will be a good future optimization, but there are other related optimizations I might pursue first. For example, it should be possible to avoid allocating new histogram buckets for each collection, by swapping them, but this will require care and can be done later.)

Reply:
Thanks for the explanation. Two follow up questions on this:

1. Given that we retain the entries that were updated in the just completed collection cycle, doesn't that mean the comment is wrong as I originally thought, i.e. we can also get to this point if "(c) stateful aggregation is NOT being used, but the aggregator was updated in the previous collection cycle"

2. Is the reason that memory will grow unbounded because we may get different label sets over time and don't want to retain stale ones? How do we resolve that issue for stateful values?

Reply:
1. Stateless values are deleted from the map at the end of a collection.  They could be preserved for another round, to keep the map from shrinking and growing (and general allocations), in which case we'd skip those entries during `ForEach`. I considered this, but it was going to complicate the already-complicated logic, at least so I figured at the time.

2. Yes, because label sets can change over time. We have no way to address this for stateful values, which is a downside of stateful values. This is why OTLP is a pass-through exporter. Choose OTLP and you won't have stateful aggregators, thus no memory growth.

Reply:
> 1. Stateless values are deleted from the map at the end of a collection

But as per @MrAlias's comment above, stateless values are only deleted if they were not updated in the just-ran collection cycle (line [277](https://github.com/open-telemetry/opentelemetry-go/pull/856/files/80d40963f617744cdcf85ca540233852e01342f4#diff-145418ef55f350c73b23bc7f4a3fd710R277)).

==> if a labelset is updated every cycle, it will always have a value in `b.values` even if it is a stateless metric ... meaning we can reach this point in the code despite what the comment says.

> I considered this, but it was going to complicate the already-complicated logic, at least so I figured at the time.

Yea makes sense. On a related note, are we happy with the fact that we will report stale stateful aggregations forever? (I think stale is the right terminology here)

> 2. Yes, because label sets can change over time. We have no way to address this for stateful values, which is a downside of stateful values. This is why OTLP is a pass-through exporter. Choose OTLP and you won't have stateful aggregators, thus no memory growth.

👍

Reply:
> meaning we can reach this point in the code despite what the comment says

I see. Yes.

> are we happy with the fact that we will report stale stateful aggregations forever?

This is a good question.

For delta exporters, the common case would be a SumObserver and I think reporting a 0 value is appropriate, since the contract for asynchronous instruments is that [LabelSet presence is meaningful in addition to the value](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/metrics/api.md#asynchronous-observations-form-a-current-set).

For cumulative exporters, we can feel unhappy about this. But this is the behavior that Prometheus expects, as far as I understand.